src/oge/helpers.py

```diff
@@ -579,6 +614,136 @@ def add_plant_entity(df: pd.DataFrame) -> pd.DataFrame:
     return df
 
 
+def add_missing_location(df: pd.DataFrame) -> pd.DataFrame:
+    """Add missing latitude, longitude, state, county and city when possible.
+
+    Args:
+        df (pd.DataFrame): table with 'latitude', 'longitude', 'state', 'county' and
+            'city' columns
+
+    Returns:
+        pd.DataFrame: original data frame with missing 'latitude', 'longitude',
+            'state', 'county' and 'city' filled out when possible.
+    """
+    # get lat/lon
+    missing_coord = df[df["longitude"].isna() | df["latitude"].isna()]
+    if len(missing_coord) > 0:
+        # only get coordinates when state, county and city are available
+        for i in missing_coord.index:
+            state = df.loc[i, "state"]
+            county = df.loc[i, "county"]
+            city = df.loc[i, "city"]
+            # if county == "NOT IN FILE" or city == "unsited":
+            #     continue
+            # else:
+            lat, lon = get_coordinates_of_location(state, county, city)
+            df.loc[i, "latitude"] = lat
+            df.loc[i, "longitude"] = lon
+
+    # get missing state, county and city from coordinates
+    missing_location = df[df["state"].isna() | df["county"].isna() | df["city"].isna()]
+    if len(missing_location) > 0:
+        for i in missing_location.index:
+            if df.loc[i, ["latitude", "longitude"]].isna().sum() == 0:
+                state, county, city = search_location_from_coordinates(
+                    df.loc[i, "latitude"],
+                    df.loc[i, "longitude"],
+                )
+                if pd.isna(df.loc[i, "state"]):
+                    df.loc[i, "state"] = state
+                if pd.isna(df.loc[i, "county"]):
+                    df.loc[i, "county"] = county
+                if pd.isna(df.loc[i, "city"]):
+                    df.loc[i, "city"] = city
+
+    return df
+
+
+def search_location_from_coordinates(latitude: float, longitude: float) -> tuple[str]:
```

Comment:
Should we add some functionality to this to validate that the returned location is in the United States?

Reply:
Yes. I will add some validations

Reply:
Done